File: languages/nodejs/src/operations/quick-sort.ts
/**
 * Implements the QuickSort algorithm.
 */

import { Comparable } from '@utils/comparable';

function swap<T>(arr: Comparable<T>[], i: number, j: number) {
    [arr[i], arr[j]] = [arr[j], arr[i]];
}

function partition<T>(arr: Comparable<T>[], low: number, high: number): number {
    const pivot = high;     // Last element is pivot
    let i = low - 1;        // Index of smaller element

    for (let j = low; j < high; j++) {
        if (arr[j].compareTo(arr[pivot]) < 0) {
            i++;
            swap(arr, i, j);
        }
    }

    i++;
    swap(arr, i, pivot);
    return i;
}

function quickSortInner<T>(arr: Comparable<T>[], low: number, high: number): Comparable<T>[] {
    if (low < high) {
        const pivotIndex = partition(arr, low, high);
        quickSortInner(arr, low, pivotIndex - 1);
        quickSortInner(arr, pivotIndex + 1, high);
    }
    return arr;
}

export function quickSort<T>(arr: Comparable<T>[]): Comparable<T>[] {
    return quickSortInner(arr, 0, arr.length - 1);
}

File: languages/ruby/lib/operations/quick_sort.rb
# Implements the QuickSort algorithm.

module QuickSortOperation
  def self.quick_sort(arr)
    self.quick_sort_inner(arr, 0, arr.length() - 1)
  end

  private_class_method def self.partition(arr, low, high)
    pivot = high
    i = low - 1
  
    for j in low..high
      if (arr[j] <=> arr[pivot]) < 0 then
        i += 1
        arr[i], arr[j] = arr[j], arr[i]
      end
    end
  
    i += 1
    arr[i], arr[pivot] = arr[pivot], arr[i]
  
    i
  end
  
  private_class_method def self.quick_sort_inner(arr, low, high)
    if low < high then
      pivot_index = partition(arr, low, high)
      quick_sort_inner(arr, low, pivot_index - 1)
      quick_sort_inner(arr, pivot_index + 1, high)
    end
    
    arr
  end
end